### DISCUSSION TITLE
What is the ideal setup for encapsulating business logic related to Recoil Atoms?

### DISCUSSION THREAD

**Discussion by AdamJaggard** [facebookexperimental/Recoil#1316] [Q&A]
This is a problem I haven't quite figured out a best practice for but I have some recipes I use successfully on projects but they feel less than ideal for some reason. I'm also aware that Recoil is puposefully very flexible and unopinionated in this regard but this has led to lots of searching for a clean solution. I'd be very interested to see a section in the docs with recipes for handling more complex scenarios involving a lot of business logic.

For example let's say I have an atom `countState`, I have various pieces of functionality relating to the atom, `addToCount`, `doubleCount`, `divideAndRound`, etc, (simplified examples), my question is where should these functions live?

They could live at one of the top components of my app but now they need passing around as props. This is the opposite paradigm to how the state itself is consumed (imported directly from consuming components).

The pattern I have been using recently looks somewhat like this:

```javascript
const useCountFunctions = () => {
  const setCount = useSetRecoilState(countState);

  // these call setCount and do what that need to do
  const addToCount = () => {}; 
  const doubleCount = () => {};
  const divideAndRound = () => {};

  return {addToCount, doubleCount, divideAndRound};
}

const Component = (props) => {
  const count = useRecoilValue(countState);
  const { addToCount } = useCountFunctions();
}
```

This "works" (in a simple example) but each function returned is recreated on every render, and each call to `useCountFunctions` creates a new copy of that hooks scope. This means that adding in a useEffect will essentially execute the effect for every instance of `useCountFunctions` that is being used. The model here is therefore the API I've created is conceptually attached to the rendering of components, where what I conceptually wanted was an API that lived outside of the components that acts as a middle man to updating my state/atoms. Because it's not possible to just write a pure function outside of the React tree and have it update an atom for you I'm not sure on the best place to put this functionality that isn't concerned with the rendering of components, only in handling logic when updating state.

Would love to see other people's approach to this.

In my mind, I want to be able to do something like this (not a well thought out suggestion) with the implementation written in a way that it is disassociated from the rendering of the component.

```javascript
const Component = () => {
  const [count, { setCount, doubleCount, etc }] = useRecoilState(countState);
}
```

Let me know if this rambling makes any sense 🙃

**Comment by drarmstr** [comment]
It's perfectly reasonable to compose React hooks to create abstractions or normalize your functionality.  If you are concerned about unique instances of functions the setters returned by Recoil are memoized per atom/selector and `<RecoilRoot>`.  But, you can also always leverage `useCallback()` or `useMemo()` if necessary to help with memoization for unique instances.

Also, you can use hooks like `useRecoilTransaction_UNSTABLE()` or `useRecoilCallback()` to create callbacks that accept different actions or parameters to centralize your business logic.  Here's a [reducer example](https://recoiljs.org/docs/api-reference/core/useRecoilTransaction#reducer-example)

**Comment by AdamJaggard** [comment]
Hey @drarmstr, thank you for the quick reply!

I looked into using useCallback/Memo a bit more but was wary of just applying it to everything, perhaps an unwarrented concern though.

`useRecoilTransaction_UNSTABLE` looks very interesting. I'll have a play with it. It seems like it might do something along the lines of what I conceptually want to achieve (grouping business logic closer to the atoms instead of being part of the render tree).

What's the current roadmap for stabilising the `useRecoilTransaction` api? I'll keep on eye out for updates.
